src/ngraph/runtime/cpu/cpu_external_function.cpp

```diff
@@ -1209,27 +1211,117 @@ void runtime::cpu::CPU_ExternalFunction::build()
+            // Build the flow graph
+            if (first_iteration)
             {
-                for (size_t j = 0; j < p.second; j++)
+                std::unordered_map<
+                    std::string,
+                    tbb::flow::continue_node<tbb::flow::continue_msg, tbb::flow::lightweight>*>
+                    nodename_tbbnode_map;
+                tbb::flow::continue_node<tbb::flow::continue_msg,
+                                         tbb::flow::lightweight>* flowgraph_node_start =
+                    new tbb::flow::continue_node<tbb::flow::continue_msg, tbb::flow::lightweight>(
+                        *(ctx->G), [&](const tbb::flow::continue_msg& msg) {});
+                auto it = enable_nodename_list.begin();
+                for (const auto& p : enables)
+                {
+                    tbb::flow::continue_node<tbb::flow::continue_msg, tbb::flow::lightweight>*
+                        flowgraph_node = new tbb::flow::continue_node<tbb::flow::continue_msg,
+                                                                      tbb::flow::lightweight>(
+                            *(ctx->G), [&](const tbb::flow::continue_msg& msg) {
+                                if (p.first(ctx) || first_iteration)
+                                {
+                                    for (size_t j = 0; j < p.second; j++)
+                                    {
+                                        (*functor)(ctx);
+                                        std::advance(functor, 1);
+                                    }
+                                }
+                                else
+                                {
+                                    std::advance(functor, p.second);
+                                }
+                            });
+                    nodename_tbbnode_map.insert({it->second, flowgraph_node});
+                    it++;
+                }
+
+                vector<Node*> dependence_graph_heads;
+
+                traverse_nodes(
+                    m_function,
+                    [&dependence_graph_heads, &nodename_tbbnode_map](shared_ptr<Node> n) {
+                        if (!n->is_parameter() && !n->is_constant())
+                        {
+                            bool is_head = true;
+                            for (auto arg : n->get_arguments())
+                            {
+                                if (!arg->is_parameter() && !arg->is_constant())
+                                {
+                                    is_head = false;
+                                    tbb::flow::make_edge(*(nodename_tbbnode_map[arg->get_name()]),
+                                                         *(nodename_tbbnode_map[n->get_name()]));
+                                }
+                            }
+                            if (is_head)
+                            {
+                                dependence_graph_heads.emplace_back(n.get());
```

Comment:
Instead of collecting into dependence_graph_heads, and subsequently creating a flowgraph edge, directly create the edge between flowgraph_node_start and this node here. 

change codegen logic as well to not skip dependence_graph_heads.